File: src/main/java/gg/climb/lolobjects/game/Composition.java
package gg.climb.lolobjects.game;

import org.immutables.value.Value;

import java.util.Arrays;
import java.util.List;

/**
 * Represents a team composition that can be used for each side of the map. You may obtain an
 * instance through {@link CompositionBuilder}
 */
@Value.Style(visibility = Value.Style.ImplementationVisibility.PRIVATE,
             strictBuilder = true)
@Value.Immutable
public abstract class Composition {
  public abstract Champion getTop();

  public abstract Champion getJungle();

  public abstract Champion getMid();

  public abstract Champion getAdc();

  public abstract Champion getSupport();

  @Value.Derived
  public List<Champion> getChampions() {
    return Arrays.asList(getTop(), getJungle(), getMid(), getAdc(), getSupport());
  }
}

File: src/main/scala/gg/climb/lolobjects/game/Composition.scala
package gg.climb.lolobjects.game

/**
 * Represents a team composition that can be used for each side of the map.
 */
class Composition(top: Champion, jungle: Champion, mid: Champion, bot: Champion, support: Champion) {
  def list = List(top, jungle, mid, bot, support)
}
